From 62c4a300d876886bf4b5ad2ad80d5f709ad07b62 Mon Sep 17 00:00:00 2001
From: Matthew Schwartz <matthew.schwartz@linux.dev>
Date: Thu, 11 Jun 2026 10:24:42 -0700
Subject: [PATCH 133/188] hid-asus-ally: add sleep_animation attr for ring LEDs

The MCU keeps a persistent mask of power states in which it animates
the joystick rings (factory default: all of them). With the sleep bit
set, the MCU breathes the rings through s2idle whenever it has power,
which on AC is always, regardless of mcu_powersave.

Clear the sleep bit by default and expose it as a sleep_animation
attribute on the joystick_rings LED device.

For upstream, we should figure out a way to expose this functionality
in a standardized way across devices, similar to LED controls.

Closes: https://github.com/ValveSoftware/SteamOS/issues/2262
Signed-off-by: Matthew Schwartz <matthew.schwartz@linux.dev>
---
 drivers/hid/hid-asus-ally.c | 92 +++++++++++++++++++++++++++++++++++++
 drivers/hid/hid-asus-ally.h |  2 +
 2 files changed, 94 insertions(+)

diff --git a/drivers/hid/hid-asus-ally.c b/drivers/hid/hid-asus-ally.c
index dc7d5de94..1cbc86328 100644
--- a/drivers/hid/hid-asus-ally.c
+++ b/drivers/hid/hid-asus-ally.c
@@ -384,6 +384,7 @@ struct ally_rgb_dev {
 	spinlock_t lock;
 
 	bool removed;
+	bool attr_group_added;
 	bool update_rgb;
 	uint8_t red[4];
 	uint8_t green[4];
@@ -1737,6 +1738,82 @@ static int ally_rgb_set_bright_base_max(struct hid_device *hdev)
 	return asus_dev_set_report(hdev, buf, sizeof(buf));
 }
 
+/*
+ * Persistent mask of power states in which the MCU animates the joystick
+ * rings (factory default: all). With the sleep bit set the MCU breathes the
+ * rings through s2idle whenever it has power, which on AC it always does,
+ * regardless of mcu_powersave.
+ */
+#define ALLY_LED_POWER_BOOT		BIT(0)
+#define ALLY_LED_POWER_AWAKE		BIT(1)
+#define ALLY_LED_POWER_SLEEP		BIT(2)
+#define ALLY_LED_POWER_SHUTDOWN		BIT(3)
+
+/* Default to no animation during sleep, adjustable via sleep_animation attr */
+static u8 ally_led_power_mask = ALLY_LED_POWER_BOOT | ALLY_LED_POWER_AWAKE |
+				ALLY_LED_POWER_SHUTDOWN;
+
+static int ally_set_led_power_states(struct hid_device *hdev, u8 mask)
+{
+	u8 buf[16] = { [0] = FEATURE_ROG_ALLY_REPORT_ID,
+		       [1] = FEATURE_ROG_ALLY_CODE_PAGE,
+		       [2] = xpad_cmd_set_led_power,
+		       [3] = xpad_cmd_len_led_power,
+		       [4] = mask };
+
+	return asus_dev_set_report(hdev, buf, sizeof(buf));
+}
+
+static ssize_t sleep_animation_show(struct device *dev,
+				    struct device_attribute *attr, char *buf)
+{
+	return sysfs_emit(buf, "%d\n", !!(ally_led_power_mask & ALLY_LED_POWER_SLEEP));
+}
+
+static ssize_t sleep_animation_store(struct device *dev,
+				     struct device_attribute *attr,
+				     const char *buf, size_t count)
+{
+	struct led_classdev *cdev = dev_get_drvdata(dev);
+	struct led_classdev_mc *mc_cdev = lcdev_to_mccdev(cdev);
+	struct ally_rgb_dev *led_rgb = container_of(mc_cdev, struct ally_rgb_dev, led_rgb_dev);
+	bool enable;
+	int ret;
+
+	ret = kstrtobool(buf, &enable);
+	if (ret < 0)
+		return ret;
+
+	if (enable)
+		ally_led_power_mask |= ALLY_LED_POWER_SLEEP;
+	else
+		ally_led_power_mask &= ~ALLY_LED_POWER_SLEEP;
+
+	ret = ally_set_led_power_states(led_rgb->hdev, ally_led_power_mask);
+	if (ret < 0)
+		return ret;
+
+	return count;
+}
+DEVICE_ATTR_RW(sleep_animation);
+
+static ssize_t sleep_animation_index_show(struct device *dev,
+					  struct device_attribute *attr, char *buf)
+{
+	return sysfs_emit(buf, "0 1\n");
+}
+DEVICE_ATTR_RO(sleep_animation_index);
+
+static struct attribute *ally_rgb_attrs[] = {
+	&dev_attr_sleep_animation.attr,
+	&dev_attr_sleep_animation_index.attr,
+	NULL
+};
+
+static const struct attribute_group ally_rgb_attr_group = {
+	.attrs = ally_rgb_attrs,
+};
+
 static void ally_rgb_do_work(struct work_struct *work)
 {
 	struct ally_rgb_dev *led = container_of(work, struct ally_rgb_dev, work);
@@ -1926,6 +2003,19 @@ static struct ally_rgb_dev *ally_rgb_create(struct hid_device *hdev)
 
 	ally_rgb_set_bright_base_max(hdev);
 
+	/* The multicolor LED core owns led_cdev->groups, add ours separately */
+	ret = device_add_group(led_rgb->led_rgb_dev.led_cdev.dev, &ally_rgb_attr_group);
+	if (ret) {
+		hid_err(hdev, "Failed to create sleep_animation attributes: %d\n", ret);
+		goto skip_led_power;
+	}
+	led_rgb->attr_group_added = true;
+
+	ret = ally_set_led_power_states(hdev, ally_led_power_mask);
+	if (ret < 0)
+		hid_err(hdev, "Failed to set ring LED power states: %d\n", ret);
+
+skip_led_power:
 	/* Not marked as initialized unless ally_rgb_set() is called */
 	if (drvdata.led_rgb_data.initialized) {
 		msleep(1500);
@@ -1954,6 +2044,8 @@ static void ally_rgb_remove(struct hid_device *hdev)
 	led_rgb->output_worker_initialized = false;
 	spin_unlock_irqrestore(&led_rgb->lock, flags);
 	cancel_work_sync(&led_rgb->work);
+	if (led_rgb->attr_group_added)
+		device_remove_group(led_rgb->led_rgb_dev.led_cdev.dev, &ally_rgb_attr_group);
 	devm_led_classdev_multicolor_unregister(&hdev->dev, &led_rgb->led_rgb_dev);
 
 	hid_info(hdev, "Removed Ally RGB interface");
diff --git a/drivers/hid/hid-asus-ally.h b/drivers/hid/hid-asus-ally.h
index c83817589..c73e11619 100644
--- a/drivers/hid/hid-asus-ally.h
+++ b/drivers/hid/hid-asus-ally.h
@@ -26,6 +26,7 @@ enum xpad_cmd {
 	xpad_cmd_set_tr_dz = 0x05, /* deadzones */
 	xpad_cmd_set_vibe_intensity = 0x06,
 	xpad_cmd_set_leds = 0x08,
+	xpad_cmd_set_led_power = 0x09,
 	xpad_cmd_check_ready = 0x0A,
 	xpad_cmd_set_turbo = 0x0F,
 	xpad_cmd_set_response_curve = 0x13,
@@ -39,6 +40,7 @@ enum xpad_cmd_len {
 	xpad_cmd_len_deadzone = 0x04,
 	xpad_cmd_len_vibe_intensity = 0x02,
 	xpad_cmd_len_leds = 0x0C,
+	xpad_cmd_len_led_power = 0x01,
 	xpad_cmd_len_turbo = 0x20,
 	xpad_cmd_len_response_curve = 0x09,
 	xpad_cmd_len_adz = 0x02,
-- 
2.55.0

